### PR TITLE
Demanded bits with fixes 

### PR DESCRIPTION
**Pull request by pranavk**
This is a takeover of PR #639 with minor fixes on top of it.

### CODE REVIEW

File: lib/Infer/AbstractInterpreter.cpp [hunk 1]
```diff
@@ -968,7 +967,83 @@ namespace souper {
...
+
+  std::unordered_map<Inst *, llvm::APInt>
+          MustDemandedBitsAnalysis::findMustDemandedBits(souper::Inst* I) {
+    if (Cache.find(I) != Cache.end()) {
+      return Cache[I];
+    }
+    std::unordered_map<Inst *, llvm::APInt> Result;
+    switch (I->K) {
+      case Inst::Var:
+        Result[I] = llvm::APInt(I->Width, -1);
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: as I commented on the original PR, this almost certainly isn't correct -- make sure you've addressed all comments on the original PR

File: include/souper/Infer/AbstractInterpreter.h [hunk 2]
```diff
@@ -122,11 +122,70 @@ namespace souper {
                                                             std::vector<InstMapping> &PCs);
   };
 
+  struct UseMap {
+    std::set<Inst *>
+    operator() (Inst *I) {
```
Comment (regehr) [inline include/souper/Infer/AbstractInterpreter.h]: fix indentation

Comment (regehr) [reply include/souper/Infer/AbstractInterpreter.h]: also must we overload operator()? why?

File: include/souper/Infer/AbstractInterpreter.h [hunk 3]
```diff
@@ -122,11 +122,70 @@ namespace souper {
...
+      return Map[I];
+    }
+
+    void compute(Inst *I) {
+      if (Map.find(I) != Map.end()) return;
+
+      std::set<Inst *> Accumulator;
+      for (auto Op : I->Ops) {
+        compute(Op);
+        for (auto V : Map[Op]) Accumulator.insert(V);
```
Comment (regehr) [inline include/souper/Infer/AbstractInterpreter.h]: put body of a loop on the next line

File: lib/Infer/AbstractInterpreter.cpp [hunk 4]
```diff
@@ -974,7 +973,83 @@ namespace souper {
...
+      default:
+        break;
+    }
+    return Result;
+  }
+#undef MDB0
+#undef MDB1
+#undef MDB2
+
+  std::unordered_map<Inst *, llvm::APInt>
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: a typedef will help with readability

File: lib/Infer/Pruning.cpp [hunk 5]
```diff
@@ -169,6 +169,26 @@ bool PruningManager::isInfeasible(souper::Inst *RHS,
     }
   }
 
+  if (!HasHole) {
+    auto MayDemandedBits = DontCareBitsAnalysis().findDontCareBits(RHS);
```
Comment (regehr) [inline lib/Infer/Pruning.cpp]: plz rename, let's not refer to MayDemanded at all

File: lib/Infer/Pruning.cpp [hunk 6]
```diff
@@ -169,6 +169,26 @@ bool PruningManager::isInfeasible(souper::Inst *RHS,
...
+
+    for (auto Pair : LHSMustDemandedBits) {
+      assert(MayDemandedBits.find(Pair.first) != MayDemandedBits.end() && "RHS Demanded bits info missing.");
+
+      if ((Pair.second & ~MayDemandedBits[Pair.first]) != 0) {
+        // This input is must demanded in lhs and NOT may demanded in rhs
+        llvm::errs() << "Var : " << Pair.first->Name << " : ";
```
Comment (regehr) [inline lib/Infer/Pruning.cpp]: why is this printing unconditionally? doesn't seem like what we want

**Comment by regehr** [comment]
needs a few test cases, both positive and negative

**Comment by manasij7479** [comment]
@googlebot I consent.

File: include/souper/Util/LLVMUtils.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+#ifndef SOUPER_UTIL_LLVMUTILS_H
+#define SOUPER_UTIL_LLVMUTILS_H
+
+namespace souper {
+
+  std::string inline getPaddedBinaryString(llvm::APInt Result) {
+    auto W = Result.getBitWidth();
+    auto Str = Result.toString(2, false);
+    if (W > Str.length()) {
```
Comment (regehr) [inline include/souper/Util/LLVMUtils.h]: please delete this redundant test

File: lib/Infer/AbstractInterpreter.cpp [hunk 8]
```diff
@@ -901,7 +902,6 @@ namespace souper {
       case Inst::Xor :
         Result = RB0 & RB1;
         break;
-
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: please avoid spurious whitespace changes

File: lib/Infer/AbstractInterpreter.cpp [hunk 9]
```diff
@@ -920,7 +920,6 @@ namespace souper {
       case Inst::Sub :
         Result = RB0 & RB1;
         break;
-
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: same

File: lib/Infer/AbstractInterpreter.cpp [hunk 10]
```diff
@@ -983,7 +982,106 @@ namespace souper {
...
+
+    for (auto V : Inputs) {
+      Result[V] = llvm::APInt::getNullValue(V->Width);
+    }
+
+    return Result;
+  }
+
+}
+
```
Comment (regehr) [inline lib/Infer/AbstractInterpreter.cpp]: kill blank line at end of file

**Comment by regehr** [comment]
nice test cases

**Comment by pranavk** [comment]
updated PR with comments resolved.

**Comment by regehr** [comment]
LGTM
